### ISSUE TITLE
Timing issue causes p1 to stop working

### ISSUE THREAD

**Issue by tomsteenbakkers** [ebaauw/homebridge-p1#12]
After updating to the new version 0.1.10 the plugin is not working anymore. Plugin is loading but after homebridge is started P1 stopped and the HomeKit accessories is not visible.

This is a past of the log:

```
[9/13/2018, 8:21:46 PM] Homebridge is running on port 51826.
[9/13/2018, 8:21:51 PM] [Yeelight] all known devices found, stopping proactive search.
[9/13/2018, 8:21:58 PM] [P1] XMX5LGBBFFB231110695 v4.2
[9/13/2018, 8:21:58 PM] [P1] Error: This callback function has already been called by someone else; it can only be called one time.
    at /usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/lib/util/once.js:12:13
    at P1Client.p1.on (/usr/local/lib/node_modules/homebridge-p1/lib/P1Platform.js:132:11)
    at P1Client.emit (events.js:182:13)
    at P1Client.parseData (/usr/local/lib/node_modules/homebridge-p1/lib/P1Client.js:402:12)
    at ReadLineParser.parser.on (/usr/local/lib/node_modules/homebridge-p1/lib/P1Client.js:280:46)
    at ReadLineParser.emit (events.js:182:13)
    at addChunk (_stream_readable.js:283:12)
    at readableAddChunk (_stream_readable.js:260:13)
    at ReadLineParser.Readable.push (_stream_readable.js:219:10)
    at ReadLineParser.Transform.push (_stream_transform.js:151:32)
[9/13/2018, 8:22:08 PM] [P1] Electricity: set homekit total consumption from 21692.46 kWh to 21692.467 kWh
[9/13/2018, 8:22:08 PM] [P1] Electricity: set homekit current consumption from 2262 W to 2241 W
[9/13/2018, 8:22:18 PM] [P1] Electricity: set homekit total consumption from 21692.467 kWh to 21692.473 kWh
[9/13/2018, 8:22:18 PM] [P1] Electricity: set homekit current consumption from 2241 W to 2247 W
[9/13/2018, 8:22:28 PM] [P1] Electricity: set homekit total consumption from 21692.473 kWh to 21692.479 kWh
```

**Comment by ebaauw** [comment]
Did you update from v1.0.9 or from an earlier version?

**Comment by tomsteenbakkers** [comment]
from v1.0.9. During startup I also see the following message

`[9/13/2018, 8:21:46 PM] [P1] /dev/ttyUSB0: no valid data received`

**Comment by ebaauw** [comment]
Ah, homebridge-zp timed out on reading the first telegram from the P1 port.  The error handling calls the callback function with an empty accessory list, so homebridge won't show the P1 accessories.  I assume the other accessories (Yeelight, other plugins) are working?  When the telegram arrives later on, homebridge-p1 illegally calls the callback function again, now with the accessories.  This is a bug in the logic.

I don't think it's related to v0.1.10, just to unlucky timing.  I see you have a DSMR v4.2 meter.  Does that issue a telegram every second, or less frequently, like every 10 seconds?  In that case, could you try and set `"timeout": 50` in config.json.

Would you be able to run homebridge -D and capture the debug message with the telegram.  I haven't seen a 4.2 telegram before.

**Comment by tomsteenbakkers** [comment]
Added the parameter timeout tp 50 and now it’s working again. I see that from September 3 now data is registered in HomeKit. So I think this has to to with the upgrade to the latest version of homebridge. 

Is it possible to get the data from the last 10 days?

**Comment by ebaauw** [comment]
If homebridge-p1 was running, the data is stored in `~/.homebridge/accessories/history_*.json`.  It might take Eve some time to load the missing data into the app, but it should become available.  If homebridge-p1 wasn't running (same timeout error as after the homebridge-p1 upgrade), there's no data to load.

**Comment by tomsteenbakkers** [comment]
ATA is loading. Was a little impatient. Last questions. What is the parameter timeout doing?

**Comment by ebaauw** [comment]
I'm still using the static accessory model for my plugin, where the plugin reports the accessories to be exposed to homebridge on startup.  If the plugin fails to report this, homebridge hangs and doesn't start the server to which the HomeKit clients connect.  To prevent this. I set a timeout for discovering the accessories.  The `"timeout"` parameter in config.json specifies how long before this timeout occurs.  It should be long enough for homebridge-p1 to receive a full telegram, so it knows what accessories (electricity, gas) it should expose and with what services and characteristics.

The parser for the telegrams only recognises the full telegram at the start of the next telegram.  For DSMR 5.0 the meter sends a telegram every second, so the first full telegram should be recognised  within two seconds.  If the meter sends a telegram only every 10 seconds (which is the case for DSMR 2.2 - I don't know about DSMR 4.2), this might take up to 20 seconds.  I've added some margin for a communication glitch (checksum error) or another plugin doing synchronous I/O.
